### PR TITLE
WIP: Use avx when available.

### PR DESCRIPTION
**Pull request by Dooez**
Added runtime detection of instruction set copied from [Mysticial](https://github.com/Mysticial/FeatureDetector).
Added two additional `scan_for_text_ascii` with bigger simd vector sizes and compile flags to enable the compilation.
`-mavx512bitalg`  may be unnecessary for 512bit version.

### CODE REVIEW

File: src/libunicode/CMakeLists.txt [hunk 1]
```diff
@@ -234,7 +254,7 @@ if(LIBUNICODE_TESTING)
     if(NOT Catch2_FOUND)
         # supress conversion warnings for Catch2
         # https://github.com/catchorg/Catch2/issues/2583
-        # https://github.com/SFML/SFML/blob/e45628e2ebc5843baa3739781276fa85a54d4653/test/CMakeLists.txt#L18-L22
+        # https://github.com/SFML/SFML/blob/e45628e2ebc5843baa3739781276fa85a54d4653/test/CMakeLists.txt #L18-L22
```
Comment (Yaraslaut) [inline src/libunicode/CMakeLists.txt:257]: I suppose editor did not realize this is a comment

File: src/libunicode/scan.cpp [hunk 2]
```diff
@@ -22,19 +23,7 @@
...
-    namespace stdx = std::experimental;
-#elif __has_include(<simd>) && defined(LIBUNICODE_USE_STD_SIMD)
-    #define USE_STD_SIMD
-    #include <simd>
-    namespace stdx = std;
-#elif defined(__SSE2__)
-    #include <immintrin.h>
-#endif
-// clang-format on
+#include "scan_simd_impl.hpp"
```
Comment (Yaraslaut) [inline src/libunicode/scan.cpp:26]: Use `<` for include and `h` file extension

File: src/libunicode/intrinsics.h [hunk 3]
```diff
@@ -72,6 +73,144 @@ struct platform_intrinsics<__m128i>
...
+    static inline mask_t greater(vec_t a, vec_t b) noexcept;
+
+    static inline mask_t and_mask(mask_t a, mask_t b) noexcept;
+
+    static inline mask_t or_mask(mask_t a, mask_t b) noexcept;
+
+    static inline mask_t xor_mask(mask_t a, mask_t b) noexcept;
+
+    static inline auto to_underlying(mask_t a) noexcept;
+};
```
Comment (Yaraslaut) [inline src/libunicode/intrinsics.h:108]: You can ommit this things with the folllowing
```suggestion
template <size_t SimdBitWidth, typename T>
struct intrin{};
```

**Review by Yaraslaut — CHANGES_REQUESTED** [review]
Some comments, actions are failing as well

File: src/libunicode/intrinsics.h (hunk 3 above)
Comment (Dooez) [reply src/libunicode/intrinsics.h:108]: This was purely for lsp autosuggestions.

**Review by Yaraslaut — CHANGES_REQUESTED** [review]
Thanks a lot, I checked that detection is working on both linux and windows, last thing is to fix following warnings:
```
warning: no previous declaration for ‘void cpuid(int32_t*, int32_t, int32_t)’ [-Wmissing-declarations]
   29 |         void cpuid(int32_t out[4], int32_t eax, int32_t ecx)
      |              ^~~~~
 warning: no previous declaration for ‘uint64_t xgetbv(unsigned int)’ [-Wmissing-declarations]
   33 |         uint64_t xgetbv(unsigned int index)
      |                  ^~~~~~
  warning: no previous declaration for ‘bool detect_os_avx()’ [-Wmissing-declarations]
   45 | auto detect_os_avx() -> bool
      |      ^~~~~~~~~~~~~
 warning: no previous declaration for ‘bool detect_os_avx512()’ [-Wmissing-declarations]
   64 | auto detect_os_avx512() -> bool
      |      ^~~~~~~~~~~~~~~~
```

**Review by christianparpart — APPROVED** [review]
@Dooez many thanks for your contribution. We really very much appreciate it. We've just looked together with @Yaraslaut over it. It looks overall very well implemented. Many thanks again for that.

@Yaraslaut is going to add some commit on top of it to address really minor stuff and fix the Windows platform issue. :)

File: src/libunicode/scan_simd_impl.h [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    auto const vec_control = intrinsics::set1_epi8(0x20); // 0..0x1F
+    auto const vec_complex = intrinsics::set1_epi8(-128); // equals to 0x80 (0b1000'0000)
+
+    while (input < end - simd_size)
+    {
+        auto batch = intrinsics::load(input);
+        auto is_control_mask = intrinsics::less(batch, vec_control);
+        auto is_complex_mask = intrinsics::equal(intrinsics::and_vec(batch, vec_complex), vec_complex);
+        auto ctrl_or_complex_mask = intrinsics::or_mask(is_control_mask, is_complex_mask);
+        if (is_control_mask)
```
Comment (Dooez) [inline src/libunicode/scan_simd_impl.h:81]: This  should be `ctrl_or_complex_mask`. But the tests are passing.

**Review by Yaraslaut — APPROVED** [review]

**Pull request merged**